colcon_core/topological_order.py

```diff
@@ -39,14 +39,18 @@ def topological_order_packages(
         # remove dependencies on already ordered packages
         ordered_names = {descriptor.name for descriptor in ordered.keys()}
         for q in queued:
-            q.remaining_dependencies -= ordered_names
+            deps = set(filter(lambda dependency:
+                              dependency.name not in ordered_names,
+                              q.remaining_dependencies))
+            q.remaining_dependencies = deps
```

Comment:
Instead of modifying the algorithm (which doesn't need to know about versions) I would prefer changing the input to still only contain the names rather than full dependency descriptors (I think `rec_deps` above).

Reply:
That's much simpler! Could you explain the difference between PackageDecorator and PackageDescriptor? I'm concerned PackageDecorator.recursive_dependencies is a different type than PackageDescriptor.dependencies if I simplify this.

Reply:
https://github.com/colcon/colcon-core/blob/bdba48223fbd73448caf424a934faaec35e8b4ee/colcon_core/package_descriptor.py#L8-L22

https://github.com/colcon/colcon-core/blob/bdba48223fbd73448caf424a934faaec35e8b4ee/colcon_core/package_decorator.py#L16-L24